packages/pubsub/src/queue.js

```diff
@@ -0,0 +1,200 @@
+ * @param {number} options.maxRequests - Max number of concurrent requests
+ *     allowed.
+ * @param {function(messages, callback)} options.sendHandler - Function to be
+ *     called with message queue.
+ *
+ * @example
+ * var queue = new Queue({
+ *   maxBytes: Math.pow(1024, 2) * 5,
+ *   maxMessages: 1000,
+ *   maxMilliseconds: 1000,
+ *   maxRequests: 5,
+ *   send: function(messages, done) {
+ *     var reqOpts = {
+ *       topic: self.topic.name,
+ *       messages = messages.map(function(message) {
+ *         return {
+ *           data: message.data,
+ *           attributes: message.attrs
+ *         };
+ *       })
+ *     };
+ *
+ *     return self.api.Publisher.publish(reqOpts, function(err, resp) {
+ *       if (err) {
+ *         done(err);
+ *         return;
+ *       }
+ *
+ *       done(null, {
+ *         responses: resp.messageIds
+ *       })
+ *     });
+ *   }
+ * });
+ */
+function Queue(options) {
+  this.maxBytes = options.maxBytes || Math.pow(1024, 2) * 5;
+  this.maxMessages = options.maxMessages || 1000;
+  this.maxMilliseconds = options.maxMilliseconds || 1000;
+  this.sendHandler = options.send;
+
+  this.inventory = {
+    queued: new Set(),
+    inFlight: new Set()
+  };
+
+  Object.defineProperty(this.inventory, 'queueBytes', {
+    get: function() {
+      var size = 0;
+
+      this.queued.forEach(function(message) {
+        size += message.data.size;
+      });
+
+      return size;
+    }
+  });
+
+  this.intervalHandle_ = null;
+}
+
+/**
+ * Adds message to the queue.
+ *
+ * @param {*} data - The message data.
+ * @param {function} callback - Callback to be ran once message is sent.
+ *
+ * @example
+ * queue.add(1234, function(err, resp) {});
+ */
+Queue.prototype.add = function(data, callback) {
+  var hasMaxMessages = (this.inventory.queued.size + 1) >= this.maxMessages;
+  var hasMaxBytes = (this.inventory.queueBytes + data.size) >= this.maxBytes;
+
+  if (hasMaxMessages || hasMaxBytes) {
+    this.send(this.getNextMessageBatch());
+  } else if (!this.intervalHandle_) {
+    this.startSendInterval();
+  }
+
+  this.inventory.queued.add({
+    data: data,
+    callback: callback || common.util.noop
+  });
+};
+
+/**
+ * Starts the send interval.
+ *
+ * @example
+ * queue.startSendInterval();
+ */
+Queue.prototype.startSendInterval = function() {
+  if (this.intervalHandle_) {
+    return;
+  }
+
+  var self = this;
+
+  this.intervalHandle_ = setInterval(function() {
+    self.send(self.getNextMessageBatch());
+
+    if (self.inventory.queued.size === 0) {
+      clearInterval(self.intervalHandle_);
+      self.intervalHandle_ = null;
+    }
+  }, this.maxMilliseconds);
+};
+
+/**
+ * Prepares next queue of messages to be sent.
+ *
+ * @return array
+ *
+ * @example
+ * var messages = queue.getNextMessageBatch();
+ * queue.send(messages);
+ */
+Queue.prototype.getNextMessageBatch = function() {
+  var messages = Array.from(this.inventory.queued);
+
+  this.inventory.queued.clear();
```

Comment:
I would much rather the batch implementation be changed to work like Python's does, which would solve this and at least one other major bug I pointed out last week. That would mean you do not use `setInterval` at all, and rather a single `setTimeout` per batch.